File: boj/boj24416.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int f[41];
int recCount = 0;
int dpCount = 0;

int fib(int n)
{
    if ( n == 1 || n == 2)
    {
        recCount++;
        return 1;
    }
    else{
        return ( fib(n - 1) + fib(n - 2));
    }
}

int fibonacci(int n)
{
    int i = 0;
    f[1] = f[2] = 1;

    for( i=3; i < n; i++)
    {
        dpCount++;
        f[i] = f[i - 1] + f[i - 2];
    }

    return f[n];
}

int main()
{
    int N = 0;
    memset(f, 0x00, sizeof(int) * 41);

    scanf("%d", &N);

    fib(N);
    fibonacci(N);

    printf("%d %d\n", recCount, dpCount+1);

    return 0;
}

File: boj/boj24416.py
def fib(_n:int):
    if _n == 1 or _n ==2:
        fib.counter += 1
        return 1
    else:
        return ( fib(_n - 1) + fib(_n - 2))

def fibonacci(_n:int):
    f = list()
    f.append(0)
    f.append(1)
    f.append(1)

    for i in range(3, _n + 1):
        f.append(f[i -1] + f[i-2])
        fibonacci.counter += 1
    return f[_n]



N = int(input())

fib.counter = 0
fibonacci.counter = 0

fib(N)
fibonacci(N)

print(fib.counter, fibonacci.counter)

